commerce-cloud-code/app_yourshophere/cartridge/api/Promotions.js

```diff
@@ -0,0 +1,33 @@
+/**
+ * @module app_yourshophere/cartridge/api/Promotions
+ * @description API for managing costly promotions operations and caching results across the request lifecycle
+ */
+
+const PromotionMgr = require('dw/campaign/PromotionMgr');
+
+let cachedPromotionPlan = null;
+
+/**
+ * Gets the active customer promotions with caching
+ * @returns {Object} The active customer promotions
+ */
+function getActiveCustomerPromotions() {
+    if (!cachedPromotionPlan) {
+        cachedPromotionPlan = PromotionMgr.getActiveCustomerPromotions();
+    }
+    return cachedPromotionPlan;
+}
+
```

Comment:
I dont know how the promos are implemented under the hood, but we could rather leverage the API to only read the first the promo - so in case its an actual iterator we save memory
```
function getPririotizedProductPromotion(product) {
    const topPromos = getActiveCustomerPromotions().getProductPromotions(product).toArray(0,1)
    return (topPromos && topPromos.length > 0) ? topPromos[0] : null ;
}
```